### ISSUE TITLE
Math module - should we provide lcm as well?  Or just point gcd users to BigInteger if they want it?

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#18999]
Least common multiple might quickly get into "too big to fit in a normal int" territory, so maybe we should just point to the one defined in the BigInteger module.

**Comment by bradcray** [comment]
What do other math modules do?

**Comment by lydia-duncan** [comment]
C++ provides it in the same header as `gcd` (numeric).  There doesn't seem to be any mention of when the result may be too large to fit in one of the types used.  Both `lcm` and `gcd` were added at the same time (C++17)

Python provides it in the same library as `gcd` (math) but it was only added in Python 3.9 (where `gcd` was supported since 3.5)

**Comment by bradcray** [comment]
@jabraham17 pointed out that lcm is supported on bigints, which would argue for our Math module to support it as well, alongside gcd, for feature parity with bigints.  I'm not advocating we jump on this in any short-term timeframe, though.  If a user requests it, that'd be a good motivation to put it in.

**Comment by bradcray** [comment]
OK, I am now the user who found themselves wishing we supported `lcm()` in our Math module today.  :)  (context: given an `n x m` array of locales, wanting to find the lcm in order to treat it like a square array such that each locale owned an integral number of virtual locales).

**Comment by DanilaFe** [comment]
> Python provides it in the same library as gcd (math)

Notably, Python's numbers are "big" by default. Or rather, the representation auto-converts small ints to big ones where needed. So for them, this is not a fear. I'd be curious what rust does.

Unless my math classes have gone to waste and I'm wrong, the LCM's maximum value is the product of the two arguments (in the case they are coprime). Thus, a possible approach here is to add logic to `--checks` that detects if the multiplication will overflow.

[In Rust, `lcm` that overflows creates a panic.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=dc68e947035652dd54a778a4bcbeeece) That said, this is a third-party `num` create; the standard library doesn't have LCM.

**Comment by bradcray** [comment]
> the LCM's maximum value is the product of the two arguments

Of course, the `+`, `-`, `*`, and `**` operators on integers can similarly overflow in Chapel and would similarly benefit from execution-time checks.  This is something we've discussed adding since Chapel's inception but have never prioritized.  If we were to implement such checks on those operators, the implementation of `lcm()` I threw together today would automatically inherit them (since it's implemented in terms of `*`).

**Comment by lydia-duncan** [comment]
Do we have an issue on that?  It would be good to note on that issue that tests should be added for `lcm` at the same time, since they would be impacted (and thus the behavior change would be user facing)

**Comment by bradcray** [comment]
I've opened https://github.com/chapel-lang/chapel/pull/28005 which adds an LCM routine to the Math  module and a test of it.  

I'm not sure whether there's an existing issue about testing for integer overflow/wraparound—I only looked briefly and didn't find one, but didn't look very hard.
